pkg/distributor/query.go

```diff
@@ -356,23 +418,183 @@ func sameSamples(a, b []mimirpb.Sample) bool {
 	return true
 }
 
-// Build a slice of chunks, eliminating duplicates.
-// This is O(N^2) but most of the time N is small.
-func accumulateChunks(a, b []ingester_client.Chunk) []ingester_client.Chunk {
-	ret := a
-	for j := range b {
-		if !containsChunk(a, b[j]) {
-			ret = append(ret, b[j])
+type seriesChunksStream struct {
+	StreamReader *ingester_client.SeriesChunksStreamReader
+	Series       []labels.Labels
+}
+
+func mergeSeriesChunkStreams(results []ingesterQueryResult, expectedZoneCount int) []ingester_client.StreamingSeries {
+	tree := newSeriesChunkStreamsTree(results)
+	if tree == nil {
+		return nil
+	}
+
+	allSeries := []ingester_client.StreamingSeries{}
+
+	for tree.Next() {
+		nextIngester, nextSeriesFromIngester, nextSeriesIndex := tree.Winner()
+		lastSeriesIndex := len(allSeries) - 1
+
+		if len(allSeries) == 0 || labels.Compare(allSeries[lastSeriesIndex].Labels, nextSeriesFromIngester) != 0 {
+			// First time we've seen this series.
+			series := ingester_client.StreamingSeries{
+				Labels: nextSeriesFromIngester,
+				// Why expectedZoneCount? We assume each series is present exactly once in each zone.
+				Sources: make([]ingester_client.StreamingSeriesSource, 1, expectedZoneCount),
+			}
+
+			series.Sources[0] = ingester_client.StreamingSeriesSource{
+				StreamReader: nextIngester.StreamReader,
+				SeriesIndex:  nextSeriesIndex,
+			}
+
+			allSeries = append(allSeries, series)
+		} else {
+			// We've seen this series before.
+			allSeries[lastSeriesIndex].Sources = append(allSeries[lastSeriesIndex].Sources, ingester_client.StreamingSeriesSource{
+				StreamReader: nextIngester.StreamReader,
+				SeriesIndex:  nextSeriesIndex,
+			})
+		}
+	}
+
+	return allSeries
+}
+
+func newSeriesChunkStreamsTree(results []ingesterQueryResult) *seriesChunkStreamsTree {
+	nIngesters := 0
+
+	for _, r := range results {
+		if r.streamingSeries.StreamReader != nil {
+			nIngesters++
+		}
+	}
+
+	if nIngesters == 0 {
+		return nil
+	}
+
+	t := seriesChunkStreamsTree{
+		nodes: make([]seriesChunkStreamsTreeNode, nIngesters*2),
+	}
+
+	i := 0
+
+	for _, r := range results {
+		if r.streamingSeries.StreamReader != nil {
+			t.nodes[i+nIngesters].ingester = r.streamingSeries
+			t.moveNext(i + nIngesters) // Must call Next on each item so that At() has a value.
+			i++
 		}
 	}
-	return ret
+	if nIngesters > 0 {
+		t.nodes[0].index = -1 // flag to be initialized on first call to Next().
+	}
+	return &t
+}
+
+// seriesChunkStreamsTree is a loser tree used to merge sets of series from different ingesters.
+// A loser tree is a binary tree laid out such that nodes N and N+1 have parent N/2.
+// We store M leaf nodes in positions M...2M-1, and M-1 internal nodes in positions 1..M-1.
+// Node 0 is a special node, containing the winner of the contest.
+type seriesChunkStreamsTree struct {
```

Comment:
Is this tree used to basically merge the series from different ingesters? But we can't use a min heap because we also need to retain the stateful streams and at which series each stream is at?

Do you think this should be documented somehow? Reading through the loser tree takes some time 😅

Reply:
Yep, that and a loser tree is slightly more efficient than a heap. 

I've added some explanatory comments, thanks for the suggestion.